Fix #1307: Upgrade adblock stats library to use adblock-rust.

iccub opened the pull request:
For now there is no automation when it comes to updating this library, it needs to be done manually.

<!-- 
*Thank you for submitting a pull request, your contributions are greatly appreciated!*

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. 
List any dependencies that are required for this change.
-->

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-ios/issues) for my issue if one did not already exist.
- [x] My patch or PR title has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!` (or `No Bug: <message>` if no relevant ticket)
- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New files have MPL-2.0 license header.


## Test Plan:

<!-- Any useful notes for reviewer explaining how best to test and verify. -->
Browse through popular websites and see if adblock stats are updating and the app is not crashing.
Play few youtube/twitch videos as well

### Screenshots:

<!-- If your patch includes user interface changes that you would like to suggest or that you would like UX to look at, please include them here. -->


## Reviewer Checklist:

- [x] PR is linked to an issue via [Zenhub](https://www.zenhub.com/extension).
- [x] Issues are assigned to at least one epic.
- [x] Issues include necessary QA labels:
  - [x] `QA/(Yes|No)`
  - [x] `release-notes/(include|exclude)`
  - [x] `bug` / `enhancement`
- [x] Necessary security reviews have taken place.
- [x] Adequate test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable)

iccub committed:
Fix #1307: Upgrade adblock stats library to use adblock-rust.

iccub commented on Client/WebFilters/AdblockResourcesMappings.swift:
```diff
@@ -5,7 +5,14 @@
-    static let generalAdblockName = "latest"
+    static func generalAdblockName(for fileType: FileType) -> String? {
+        switch fileType {
+        case .dat: return "rs-ABPFilterParserData"
```
Some logic had to be added to handle differences between ios-json and universal-dat files, this includes different endpoints and file names

iccub commented on Client/WebFilters/ShieldStats/Adblock/Resources/ablock_rust_lib.h:
```diff
@@ -0,0 +1,96 @@
+#ifndef ADBLOCK_RUST_FFI_H
```
This file was generated, see https://github.com/brave/adblock-rust-ffi#regenerating-the-c-header

Brandon-T commented on Client/WebFilters/AdblockResourcesMappings.swift:
```diff
@@ -14,37 +21,45 @@ enum ResourceLocale: String {
+    func resourceName(for fileType: FileType) -> String {
+        var resourceId = ""
+        
         switch self {
```
Do we prefer:
```
let map = [.ar: "...", .bg: "...", .zh: "..."]
resourceId = map[fileTypye] ?? ""
```
or `switch` statement instead?

It's just a question. You don't have to change your stuff.

iccub replied:
I don't see how this code would go then, `map` is a `locale:id`  dictionary , while `fileType` is a different enum. So I can't reallty do it like that

Brandon-T commented on Client/WebFilters/ShieldStats/Adblock/AdblockRustEngine.swift:
```diff
@@ -0,0 +1,30 @@
+        var emptyPointer: UnsafeMutablePointer<Int8>?
+        let redirect = UnsafeMutablePointer(mutating: &emptyPointer)
+        
+        return engine_match(engine, requestUrl, requestHost,
```
The `redirect` pointer is passed in but the out value is never used?

Also, do you know if:

`engine_match(...., UnsafeMutablePointer(mutating: &emptyPointer))`

would would instead of creating `redirect`?

iccub replied:
yes, we don't use any of these inout values but the library requires them, that's why I put this code in a Swift wrapper

I don't understand this part, could you rephrase
```
engine_match(...., UnsafeMutablePointer(mutating: &emptyPointer))
would would instead of creating redirect?
```

I tried passing nil value there but the lib was crashing

In general, this library is used for ad blocking on desktop and Android, while we on iOS use it only for checking if a resource is blocked, that's why most of it is not utilized

Brandon-T replied:
@iccub 

```swift
func engine_match(..., ..., redirect: UnsafeMutablePointer<UnsafeMutablePointer<Int8>?>)
```

Instead of doing:
```swift
var emptyPointer: UnsafeMutablePointer<Int8>?
let redirect = UnsafeMutablePointer(mutating: &emptyPointer)

engine_match(..., ..., redirect)
```

Would this work:

```swift
var emptyPointer: UnsafeMutablePointer<Int8>?
engine_match(..., ..., UnsafeMutablePointer(mutating: &emptyPointer))
```

Essentially getting rid of `redirect` as a temporary variable. Just curious if it would compile and work.

Otherwise just leave it how it is :)

iccub replied:
it works and compiles, changing it to your version :)

iccub committed:
Remove bundled ABPFilter data.

This file will be now always downloaded from the server if not present
on the device.

jhreis commented on Client/WebFilters/AdblockerType.swift:
```diff
@@ -7,7 +7,17 @@ import Shared
+        switch self {
+        case .dat: return "https://adblock-data.s3.brave.com/4/"
+        case .json: return "https://adblock-data.s3.brave.com/ios/"
+        case .tgz: return "" // Httpse resources are not supported yet.
```
Clarified this, but wanted to say that this does _not_ impact HTTPSE stats, as we currently use a different system for this anyway.

jhreis commented on Client/WebFilters/AdblockResourcesMappings.swift:
Creative solution to get around this 😄

jumde commented:
lgtm
